font/src/ft/mod.rs

```diff
@@ -38,13 +38,22 @@ struct FixedSize {
 
 struct FallbackFont {
     pattern: Pattern,
-    hash: PatternHash,
+    id: FontID,
 }
 
 impl FallbackFont {
-    fn new(pattern: Pattern) -> FallbackFont {
-        let hash = pattern.hash();
-        Self { pattern, hash }
+    fn new(pattern: Pattern, id: FontID) -> FallbackFont {
+        Self { pattern, id }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Copy, Clone, Hash)]
+struct FontID(u32);
+
+impl FontID {
+    fn new(lhs: PatternHash, rhs: PatternHash) -> Self {
+        // XOR two hashes to get a font ID
+        Self(((lhs.0 << 1) | (lhs.0 >> 31)) ^ rhs.0)
```

Comment:
Could you provide some (~3-5) pattern hashes, just to see what they look like? I'm curious if there's any kind of pattern recognizable, that would indicate that these aren't evenly distributed.

Reply:
Do you want pattern hashes or resulted FontIDs? If you're curious on pattern.hash() thing https://gitlab.freedesktop.org/fontconfig/fontconfig/blob/master/src/fcpat.c#L601 , so basically some fast XOR hash. The idea of my computation was from it.

Reply:
I was only curious about pattern hashes.

Reply:
Sure.

```
name: monospace, size: 9, hash: 92e9
Left hash 92e9, right hash c943209b
Left hash 92e9, right hash 29bc3ae0
Left hash 92e9, right hash 4569a792
Left hash 92e9, right hash 8e1b9f9f
Left hash 92e9, right hash 2a4fac1
Left hash 92e9, right hash a5f3b374
Left hash 92e9, right hash 66eadb05
Left hash 92e9, right hash c785e2d2
Left hash 92e9, right hash d7d1d80f
Left hash 92e9, right hash 626fe89c
Left hash 92e9, right hash 1c7e06b7
Left hash 92e9, right hash 75a316c0
Left hash 92e9, right hash 8498d10e
Left hash 92e9, right hash eba29dda
name: monospace, size: 9, hash: 92e9
Left hash 92e9, right hash c943209b
Left hash 92e9, right hash 29bc3ae0
Left hash 92e9, right hash 4569a792
Left hash 92e9, right hash 8e1b9f9f
Left hash 92e9, right hash 2a4fac1
Left hash 92e9, right hash a5f3b374
Left hash 92e9, right hash 66eadb05
Left hash 92e9, right hash c785e2d2
Left hash 92e9, right hash d7d1d80f
Left hash 92e9, right hash 626fe89c
Left hash 92e9, right hash 1c7e06b7
Left hash 92e9, right hash 75a316c0
Left hash 92e9, right hash 8498d10e
Left hash 92e9, right hash eba29dda
name: monospace, size: 9, hash: 92e9
Left hash 92e9, right hash f648c6c5
Left hash 92e9, right hash 741a66ba
Left hash 92e9, right hash 369ef7f4
Left hash 92e9, right hash 8eba312e
Left hash 92e9, right hash b18844d5
Left hash 92e9, right hash 7d572b94
Left hash 92e9, right hash 2948939d
Left hash 92e9, right hash cae4ad28
Left hash 92e9, right hash f3d4b5c8
Left hash 92e9, right hash eba29dda
Left hash 92e9, right hash 3aa7f96d
Left hash 92e9, right hash ae59d18f
Left hash 92e9, right hash a01d91d6
Left hash 92e9, right hash 87097b8d
Left hash 92e9, right hash ef8cb96a
name: monospace, size: 9, hash: 92e9
Left hash 92e9, right hash 20357cbb
Left hash 92e9, right hash 651a66ba
Left hash 92e9, right hash 8e1b9f9f
Left hash 92e9, right hash 644925a6
Left hash 92e9, right hash f76914d1
Left hash 92e9, right hash 67d1bd45
Left hash 92e9, right hash d394cf45
Left hash 92e9, right hash 4470361e
Left hash 92e9, right hash 3bad9911
Left hash 92e9, right hash e4f2d942
Left hash 92e9, right hash e65e85ad
Left hash 92e9, right hash 7751da1b
Left hash 92e9, right hash 18aa4c9c
Left hash 92e9, right hash 506d304d
name: monospace, size: 9, hash: 92e9
Left hash 92e9, right hash 27c79c84
Left hash 92e9, right hash 8756de0e
Left hash 92e9, right hash 2c9591ba
Left hash 92e9, right hash db3b3546
Left hash 92e9, right hash b18844d5
Left hash 92e9, right hash 9eb05b43
Left hash 92e9, right hash aa391d55
Left hash 92e9, right hash e098f591
Left hash 92e9, right hash b6964309
Left hash 92e9, right hash d5fca1bb
Left hash 92e9, right hash 6451a6c8
Left hash 92e9, right hash daf09927
Left hash 92e9, right hash df2096d9
Left hash 92e9, right hash a4e781b4
Left hash 92e9, right hash 85893641
Left hash 92e9, right hash 506d304d
name: monospace, size: 18.5, hash: 92f2
Left hash 92f2, right hash c94ea09b
Left hash 92f2, right hash 29bc3ae0
Left hash 92f2, right hash 4572a792
Left hash 92f2, right hash 8e009f9f
Left hash 92f2, right hash 2a97ac1
Left hash 92f2, right hash a5e8b374
Left hash 92f2, right hash 66f1db05
Left hash 92f2, right hash c79ee2d2
Left hash 92f2, right hash d7cad80f
Left hash 92f2, right hash 6274e89c
Left hash 92f2, right hash 1c6506b7
Left hash 92f2, right hash 75b816c0
Left hash 92f2, right hash 8483d10e
Left hash 92f2, right hash ebb99dda
name: monospace, size: 18.5, hash: 92f2
Left hash 92f2, right hash f64546c5
Left hash 92f2, right hash 741a66ba
Left hash 92f2, right hash 3685f7f4
Left hash 92f2, right hash 8e8c312e
Left hash 92f2, right hash b1be44d5
Left hash 92f2, right hash 7d5aab94
Left hash 92f2, right hash 2953939d
Left hash 92f2, right hash caffad28
Left hash 92f2, right hash f3cfb5c8
Left hash 92f2, right hash ebb99dda
Left hash 92f2, right hash 3abcf96d
Left hash 92f2, right hash ae42d18f
Left hash 92f2, right hash a00691d6
Left hash 92f2, right hash 873f7b8d
Left hash 92f2, right hash efbab96a
name: monospace, size: 18.5, hash: 92f2
Left hash 92f2, right hash 2038fcbb
Left hash 92f2, right hash 651a66ba
Left hash 92f2, right hash 8e009f9f
Left hash 92f2, right hash 642525a6
Left hash 92f2, right hash f75f14d1
Left hash 92f2, right hash 67cabd45
Left hash 92f2, right hash d3f8cf45
Left hash 92f2, right hash 441c361e
Left hash 92f2, right hash 3bc19911
Left hash 92f2, right hash e49ed942
Left hash 92f2, right hash e63285ad
Left hash 92f2, right hash 773dda1b
Left hash 92f2, right hash 18c64c9c
Left hash 92f2, right hash 5001304d
name: monospace, size: 18.5, hash: 92f2
Left hash 92f2, right hash 27ca1c84
Left hash 92f2, right hash 8756de0e
Left hash 92f2, right hash 2c8e91ba
Left hash 92f2, right hash dbe33546
Left hash 92f2, right hash b1be44d5
Left hash 92f2, right hash 9ebddb43
Left hash 92f2, right hash aa551d55
Left hash 92f2, right hash e083f591
Left hash 92f2, right hash b64e4309
Left hash 92f2, right hash d524a1bb
Left hash 92f2, right hash 6489a6c8
Left hash 92f2, right hash da289927
Left hash 92f2, right hash dff896d9
Left hash 92f2, right hash a43f81b4
Left hash 92f2, right hash 85513641
Left hash 92f2, right hash 5001304d

```